### PR TITLE
Feat(api): add endpoint with complex custom type

### LINKED ISSUE #32804: Handle command/query with constructor/setters using DecimalNumber and/or DateTime
Kind of related to https://github.com/PrestaShop/PrestaShop/issues/32803 because the solution will use the same service but can be started in parallel

Most of the commands parameters are scalar (string; int, bool) but some can be a bit more advanced like DecimalNumber and/or DateTime. Like in the mentioned issue, this feature of prefill/building commands should be handled by a dedicated and unit tested service. We need to handle the conversion of this kind of input, maybe ApiPlatform can handle dates natively as long as they are used in the DTO, but for `DecimalNumber` probably not.

Example:

https://github.com/PrestaShop/PrestaShop/blob/f5cfdf3fffbd1fdf967c49e933e62fed09a1d23d/src/Core/Domain/Customer/Group/Command/AddCustomerGroupCommand.php#L70

### PR DESCRIPTION
**Pull request by tleon**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop 
| Description?      | Add a new query that demonstrate how to handle custom objects as a parameter in api platform
| Type?             |  new feature 
| Category?         |  BO 
| BC breaks?        |  no
| Deprecations?     | no
| How to test?      | <ul><li>Get an bearer token by requesting the /api/oauth2/token endpoint</li><li> Create a new customer group by requesting this endpoint `/customers/group` </li></ul>
| Fixed ticket?     | Fixes #32804
| Related PRs       | N/A
| Sponsor company   | Prestashop 

More informations on how to get your bearer token :

You need to add a return statement at the start of the onKernelRequest method of the SslMiddleware.php file. If you don't do this postman won't be able to handle the current ssl protocol.
You need to modify the security_dev.yml and copy the user key of the oauth2 node from the security_test.yml

More informations about the post : 
* You need to request the endpoint with a payload looking like the following : 
```
{
  "localizedNames": [
      "toto",
      "tata"
  ],
  "reductionPercent": {
    "number": "10",
    "exponent": 1
  },
  "displayPriceTaxExcluded": true,
  "showPrice": true,
  "shopIds": [1]
}
```

### CODE REVIEW

File: src/PrestaShopBundle/ApiPlatform/Resources/CustomerGroup.php
Comment (tleon) [inline src/PrestaShopBundle/ApiPlatform/Resources/CustomerGroup.php]: This is not working. Apparently there is currently no simple way to overide the example of a nested object.
It should be possible to tinker with the serialization groups for this resource in order to impact the example of the documentation.

currently the documentation shows 
```
  "reductionPercent": {
    "exponent": 0
  },
```
It should show 
```
  "reductionPercent": {
    "number": "string",
    "exponent": "int"
  },
  ```

Comment (matks) [reply src/PrestaShopBundle/ApiPlatform/Resources/CustomerGroup.php]: So what do we do? Do we review this PR or is it faulty and you will continue to work on it?

Comment (matks) [reply src/PrestaShopBundle/ApiPlatform/Resources/CustomerGroup.php]: @tleon 🆙

**Review by matks — CHANGES_REQUESTED** [review]
Waiting for @tleon answer :p

File: tests/Integration/ApiPlatform/PostCustomerTest.php
Comment (jolelievre) [inline tests/Integration/ApiPlatform/PostCustomerTest.php]: The purpose in this case is not to define each parameter of the decimal number The API client itself would provide a value like `reductionPercent => '10.000'` and this string value must then be injected into the related CQRS command BUT it needs to be converted into a `DecimalNumber` first

Ideally, this should be done in a "generic way" meaning not only work with `DecimalNumber` but with other simple objects like VOs So during the normalization (either handled by Symfony serializer, or by a custom conversion on our side, or maybe by adding a new normalizer in the serializer) we need to detect that the command has a parameter which is not scalar but is an object and we try a simple construction by providing the value from the request as the constructor parameter

Very simple idea of the expected solution based on reflection:

```php
if (is_object($parameterType)) {
    $commandParameterValue = new $parameterType($queryValue);
}
```

Comment (mflasquin) [reply tests/Integration/ApiPlatform/PostCustomerTest.php]: The DecimalNumber does not take only one parameter so the best solution is to precise arguments in json array like describe by @tleon. I've update the CommandProcessor to use the serializer.

**Comment by mflasquin** [comment]
Tests : https://github.com/mflasquin/ga.tests.ui.pr/actions/runs/5834700689

File: tests/Integration/ApiPlatform/PostCustomerTest.php
Comment (jolelievre) [reply tests/Integration/ApiPlatform/PostCustomerTest.php]: But we never change the exponent parameter, and this second parameter is optional and already as a value of one SPlitting this into two fields only complicates the API contracts for all the decimal value we'll be using I don't think it's really adapted

File: src/PrestaShopBundle/ApiPlatform/Processor/CommandProcessor.php
Comment (matks) [inline src/PrestaShopBundle/ApiPlatform/Processor/CommandProcessor.php:40]: Moving the parameter from explicitely declared variable to constructor-declared parameter, and readonly, is a BC Break. Please document it in your PR description.

File: src/PrestaShopBundle/ApiPlatform/Processor/CommandProcessor.php
Comment (matks) [inline src/PrestaShopBundle/ApiPlatform/Processor/CommandProcessor.php]:
```suggestion
            throw new NoExtraPropertiesFoundException('Extra property "command" not found');
```

File: src/PrestaShopBundle/ApiPlatform/Provider/QueryProvider.php
Comment (matks) [inline src/PrestaShopBundle/ApiPlatform/Provider/QueryProvider.php]: Is it useful to mark it can return quite everything? 😛

File: src/PrestaShopBundle/ApiPlatform/Provider/QueryProvider.php
Comment (matks) [inline src/PrestaShopBundle/ApiPlatform/Provider/QueryProvider.php]:
```suggestion
            throw new NoExtraPropertiesFoundException('Extra property "query" not found');
```

File: tests/Integration/ApiPlatform/PostCustomerTest.php
Comment (matks) [reply tests/Integration/ApiPlatform/PostCustomerTest.php]: I agree with @jolelievre .

If I understand correctly : if we accept this implementation, it means that for _every endpoint_ where we expect a DecimalNumber, we will require the API user to provide a payload with both parameters `number` and `exponent` right? Although `exponent` should always be 1 and is useless for end-user.

We need to find a way to change this. We need that, when the Serializer processes this payload, it is able to automatically convert the number to a DecimalNumber with exponent 1.

As a end user if I want to use the API to set a percent-value, I expect to send a payload with something like "10%" or "10" or "10.0" . Not `{number:10,exponent:1}`.

**Review by matks — CHANGES_REQUESTED** [review]
@tleon Little changes requested.

Most important change is the topic raised by @jolelievre : we cannot request the API end-users to provide the `exponent` parameter. We need to find a way to tell PrestaShop Backoffice API that `exponent` will always be 1.

File: src/PrestaShopBundle/ApiPlatform/Processor/CommandProcessor.php
Comment (mflasquin) [reply src/PrestaShopBundle/ApiPlatform/Processor/CommandProcessor.php:40]: I think it's not necessary because this class exist only on develop and new api is an experimental feature

File: src/PrestaShopBundle/ApiPlatform/Provider/QueryProvider.php
Comment (mflasquin) [reply src/PrestaShopBundle/ApiPlatform/Provider/QueryProvider.php]: return mixed will be sufficient :D

File: tests/Integration/ApiPlatform/PostCustomerTest.php
Comment (mflasquin) [reply tests/Integration/ApiPlatform/PostCustomerTest.php]: It's possible to send the payload without exponent, like this :

`{
    "localizedNames": ["toto", "tata"],
    "reductionPercent": {"number": "0"},
    "displayPriceTaxExcluded": false,
    "showPrice": true,
    "shopIds": [1]
}`

WDYT ? @jolelievre @matks 
And if it's necessary we can add the exponent parameter.

File: src/PrestaShopBundle/ApiPlatform/Processor/CommandProcessor.php
Comment (matks) [reply src/PrestaShopBundle/ApiPlatform/Processor/CommandProcessor.php:40]: All right then 👍

File: tests/Integration/ApiPlatform/PostCustomerTest.php
Comment (matks) [reply tests/Integration/ApiPlatform/PostCustomerTest.php]: OK for me 👍 . If we don't burden the user with the `exponent` useless parameter, then it's good

**Review by matks — DISMISSED** [review]
Good for me

**Review by boherm — DISMISSED** [review]
LGTM, thanks @tleon!

**Review by jolelievre — CHANGES_REQUESTED** [review]
Hi @tleon 

sorry but it is still not the expected behaviour, we don't want to split a decimal number into a sub structure `decimalValue: {number; 10.6, exp: 0}` Even if the exp part is optional, the purpose of those fields is only to improve the float accuracy it doesn't mean the API should split the input

We want to provide `decimalNumber: 15.456` via the API (or `decimalNumber: '15.456'` if we want to provide even more decimals maybe), but anyway only one field that is then transformed into a DecimalNumber internally

**Comment by mflasquin** [comment]
I will fix the @jolelievre comment, I'm on a good way I think :

image

WDYT @jolelievre ?

File: src/PrestaShopBundle/ApiPlatform/Denormalizers/DecimalNumberDenormalizer.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Denormalizers/DecimalNumberDenormalizer.php]: Seems like `$type` is not used, but maybe we could use it later and transform this class into a more generic one that could also denormalize VOs?

File: src/PrestaShopBundle/ApiPlatform/Denormalizers/ObjectDenormalizer.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Denormalizers/ObjectDenormalizer.php]: Why do we need this extra class if it simply extends the existing Symfony normalizer? Besides the class doesn't seem to be used anywhere? Is it really useful?

File: src/PrestaShopBundle/ApiPlatform/Resources/CustomerGroup.php
Comment (mflasquin) [reply src/PrestaShopBundle/ApiPlatform/Resources/CustomerGroup.php]: Fixed !

File: tests/Integration/ApiPlatform/PostCustomerTest.php
Comment (mflasquin) [reply tests/Integration/ApiPlatform/PostCustomerTest.php]: Fixed !

File: tests/Integration/ApiPlatform/PostCustomerTest.php
Comment (matks) [inline tests/Integration/ApiPlatform/PostCustomerTest.php]:
```suggestion
                'reductionPercent' => 10.3,
```
Does it work with floats too? If yes, let's use it 😄 ?

File: src/PrestaShopBundle/ApiPlatform/Serializer.php
Comment (matks) [inline src/PrestaShopBundle/ApiPlatform/Serializer.php]: OK I see it's a Serializer... but which one is it? 😝 

Suggestion: find a name a little more accurate, that allows developers to guess what this class is doing from its name.

Here are some pretty poor suggestions I think you can find a better one 😄 : SetterSerializer, SetterBasedSerializer, PropertySetterSerializer

File: src/PrestaShopBundle/ApiPlatform/Serializer.php
Comment (matks) [inline src/PrestaShopBundle/ApiPlatform/Serializer.php]: That's some important logic... that I think is worth unit testing 😄 wdyt?

**Review by matks — COMMENTED** [review]
Hey @mflasquin I agree with the implementation, I have some suggestions to fine-tune the result

File: src/PrestaShopBundle/ApiPlatform/Serializer.php
Comment (mflasquin) [reply src/PrestaShopBundle/ApiPlatform/Serializer.php]: I think the namespace permit to understand for what It used, no ?

File: src/PrestaShopBundle/ApiPlatform/Serializer.php
Comment (mflasquin) [reply src/PrestaShopBundle/ApiPlatform/Serializer.php]: For the moment integrations tests are used to check if all works fine

File: tests/Integration/ApiPlatform/PostCustomerTest.php
Comment (mflasquin) [reply tests/Integration/ApiPlatform/PostCustomerTest.php]: yep :)

**Review by boherm — DISMISSED** [review]
LGTM

**Review by mflasquin — DISMISSED** [review]

**Comment by mflasquin** [comment]
tests : https://github.com/mflasquin/ga.tests.ui.pr/actions/runs/6000947900

File: src/PrestaShopBundle/ApiPlatform/Serializer.php
Comment (jolelievre) [reply src/PrestaShopBundle/ApiPlatform/Serializer.php]: Maybe `CommandSerializer` to give a bit more context and hints that the purpose is to transform API inputs into CQRS commands?

File: src/PrestaShopBundle/ApiPlatform/Serializer.php
Comment (jolelievre) [reply src/PrestaShopBundle/ApiPlatform/Serializer.php]: And yet I've been asking for this independent service and unit tests for weeks now... 😅 Which would have saved us trouble and time IMHO if they had been done from the start We keep using the excuse of lack of time and integration tests covering us every time but in the end I feel like it's making us lose more time than what's won And we definitely lose stability and trust in our code

Anyway I know you're in a rush because of the lack of time, so I won't block the PR for this, but we should try to add the unit tests by the end of the week for this service This would also mean less integration tests potentially since they are also handled via UI tests so double check of the same feature is not the most efficient way to handle all this

File: src/PrestaShopBundle/ApiPlatform/Denormalizers/DecimalNumberDenormalizer.php
Comment (jolelievre) [reply src/PrestaShopBundle/ApiPlatform/Denormalizers/DecimalNumberDenormalizer.php]: Also I wonder if we should use `Normalizer` as the namespace here, like Symfony does Because normalization and renormalization are part of the same concept and if we start introducing our custom denormalizers there are a lot of chances that at some point we'd need the opposite action (normalize) so the `Normalizer` namespace would be more generic and adapted maybe

File: tests/Integration/ApiPlatform/PostCustomerTest.php
Comment (jolelievre) [inline tests/Integration/ApiPlatform/PostCustomerTest.php:68]: I think we should also check that the entity has correctly been created since this is an integration tests Since we don't have a read endpoint available this can be done "manually" by directly checking the DB content

File: src/PrestaShopBundle/ApiPlatform/Denormalizers/ObjectDenormalizer.php
Comment (jolelievre) [reply src/PrestaShopBundle/ApiPlatform/Denormalizers/ObjectDenormalizer.php]: Oh wait I think it's used by the DI you added that automatically adds the tag on this namespace so that it can then be injected in the API serializer right?

Then you need to add a comment to explain this because this is not intuitive at all and this could be cleaned by another developer later, also you should use this class in the `QueryProviderTest` because you manually inject the symfony class not this custom additional one which is not really consistent

**Review by jolelievre — CHANGES_REQUESTED** [review]
Thanks @tleon and @mflasquin 

There are a few things to improve but the concept is interesting and promising, good job!!

File: src/PrestaShopBundle/ApiPlatform/Serializer.php
Comment (mflasquin) [reply src/PrestaShopBundle/ApiPlatform/Serializer.php]: resolved

**Review by mflasquin — DISMISSED** [review]

File: src/PrestaShopBundle/ApiPlatform/Normalizer/ObjectDenormalizer.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Normalizer/ObjectDenormalizer.php]: You can even use a negative value to be sure, which would be safer because I'm sure if the default priority is 0 or 1 🤔

File: src/PrestaShopBundle/ApiPlatform/Processor/CommandProcessor.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Processor/CommandProcessor.php]:
```suggestion
        if (null === $commandClass || !class_exists($commandClass)) {
```

Suggestion for later we could check if the provided class exist

File: src/PrestaShopBundle/ApiPlatform/Provider/QueryProvider.php
Comment (jolelievre) [inline src/PrestaShopBundle/ApiPlatform/Provider/QueryProvider.php:67]: We'll probably have to handle the merge of these parameters in the command processor later as well no?

File: tests/Integration/ApiPlatform/PostCustomerTest.php
Comment (jolelievre) [inline tests/Integration/ApiPlatform/PostCustomerTest.php:41]: Restoration/cleaning should also be done in the `tearDownAfterClass` method, the famous `leave the place as clean as you found it` 😛

File: tests/Unit/PrestaShopBundle/ApiPlatform/DomainSerializerTest.php
Comment (jolelievre) [inline tests/Unit/PrestaShopBundle/ApiPlatform/DomainSerializerTest.php]: Do you really need the `MockObject` ? Seems like it's never mocked, and it shouldn't since it's the class being tested

File: tests/Unit/PrestaShopBundle/ApiPlatform/DomainSerializerTest.php
Comment (jolelievre) [inline tests/Unit/PrestaShopBundle/ApiPlatform/DomainSerializerTest.php]: You should also provide a third parameter with the expected content of the normalized object like:
```suggestion
            AddCustomerGroupCommand::class,
            new AddCustomerGroupCommand(
                [1 => 'test1', 2 => 'test2']
                10.3,
                true,
                true,
                [1]
            )
```

Comment (jolelievre) [reply tests/Unit/PrestaShopBundle/ApiPlatform/DomainSerializerTest.php]: Of course if you do that the second parameter becomes kind od obsolete since you can get the class via a `get_class` and you could even simply use a single `assertEquals` since it checks the type internally

Nut bottom line, the most important is not just the type of the returned object but also that it's sub-fields are correctly filled

File: tests/Unit/PrestaShopBundle/ApiPlatform/StateProvider/QueryProviderTest.php
Comment (jolelievre) [inline tests/Unit/PrestaShopBundle/ApiPlatform/StateProvider/QueryProviderTest.php]: SAme here not sure the mock is useful

File: tests/Unit/PrestaShopBundle/ApiPlatform/DomainSerializerTest.php
Comment (jolelievre) [inline tests/Unit/PrestaShopBundle/ApiPlatform/DomainSerializerTest.php]:
```suggestion
                    1 => 'test1',
                    2 => 'test2',
```
Just a suggestion but it represents more of what a localized array should look like, the language ID key should be different from zero

**Review by mflasquin — APPROVED** [review]

**Review by matks — APPROVED** [review]
Good for me. @jolelievre final word?

File: tests/Unit/PrestaShopBundle/ApiPlatform/DomainSerializerTest.php
Comment (jolelievre) [inline tests/Unit/PrestaShopBundle/ApiPlatform/DomainSerializerTest.php:57]: In the following PR we'll have to add tests for normalization, the normalization is used in the QueryProvider to transform query results into flatted data before it's normalized again into an API platform resource (not necessary right away)

**Review by jolelievre — APPROVED** [review]
Thanks @mflasquin 

It should be good!

**Comment by boherm** [comment]
Tested on my side! This PR works well, good job!

**Review by boherm — APPROVED** [review]

**Comment by mflasquin** [comment]
thanks all

**Pull request merged**
